# Embeddings for Transaction Data with PyTorch

Embeddings are very popular in the computer vision and natural language processing 
domain. In this blog post I want to explore how we can use a deep learning technique called 'embeddings'
to gain insights into the structure of transaction data produced by customers at grocery stores.

I will use an anonymised dataset provided by Riccardo Guidotti on his Github [page](https://github.com/riccotti/TX-Means/tree/master/dataset)
that records sales transactions of customers in an Italian supermarket chain.

In [24]:
%connect_info
%load_ext autoreload
%autoreload 1

{
  "shell_port": 53562,
  "iopub_port": 53563,
  "stdin_port": 53564,
  "control_port": 53565,
  "hb_port": 53566,
  "ip": "127.0.0.1",
  "key": "8ec0c2e4-29b4f11c2c9dae17b82c04b1",
  "transport": "tcp",
  "signature_scheme": "hmac-sha256",
  "kernel_name": ""
}

Paste the above JSON into a file, and connect with:
    $> jupyter <app> --existing <file>
or, if you are local, you can connect with just:
    $> jupyter <app> --existing kernel-edb25871-9db7-4017-a231-632eed23f88a.json
or even just:
    $> jupyter <app> --existing
if this is the most recent Jupyter kernel you have started.
The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [23]:
import src.test
import pandas as pd

pd_data = pd.read_csv(filepath_or_buffer="../data/sample_real_data.csv", dtype='category')

pd_data.head()

test2
test3


customer_id basket_id item_id
0           0         0      10
1           0         0      12
2           0         0      13
3           0         0      15
4           0         0      20

For each customer, we record the shopping basket (basket_id) and the specific items in the basket (item_id)

Before we continue, let's take a brief look at some summary statistics:

In [6]:
pd.DataFrame.describe(pd_data)

customer_id basket_id item_id
count       329169    329169  329169
unique         100      1044     234
top             76       820       4
freq         12410      3575    8516

We have a total of 100 unique customers producing about 10 baskets each with 234 unique items. Item 4 is the most 
frequent with about 8.5k occurrences.

https://yashuseth.blog/2018/07/22/pytorch-neural-network-for-tabular-data-with-categorical-embeddings/


